config/deploy/docker/lib/roll_out.rb

```diff
@@ -434,6 +459,75 @@ def _register_task!(name:, image:, cpu_shares: nil, soft_mem_limit_mb: 512, port
     self.task_definition = results.to_h.dig(:task_definition, :task_definition_arn)
   end
 
+  # Abstraction that lets the cluster provision more/less EC2 instances based
+  # on the requirements of the containers we want to run
+  def _capacity_providers
+    @_capacity_providers ||= ecs.describe_clusters(clusters: [self.cluster]).clusters.first.capacity_providers
+  end
+
+  # https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-placement-constraints.html
+  def _placement_constraints(name)
+    [].tap do |pc|
+      # Distinct Instance is problematic if you have limited resources.
+      # pc << { type: 'distinctInstance' }
+      # puts "[INFO][CONST] constraining #{name} to distinct EC2 instances"
+
+      # Only delayed job daemons need to be on non-spot
+      if name.match?(/-dj-/)
+        puts "[INFO][CONST] Constraining #{name} to non-spot-instances (on-demand)"
+
+        pc << {
+          type: 'memberOf',
+          expression: "attribute:instance-lifecycle == #{NOT_SPOT}",
+        }
+      end
+
+      # In the interim before switching everything over to capacity providers, we only allow workoff workers
+      # to use capacity providers.
+      # FIXME: Eventually, we will remove this if-statement and replace with
+      # running DJ services with the on-demand, non-default capacity provider
+      # strategy. We need
+      #   1) for our RIs to expire
+      #   2) make the spot instance capacity provider the default one
+      #   3) redeploy everything
+      #   4) spin down the (at this point) defunct auto-scaling group to no instances
+
+      if !name.match?(/workoff/)
+        puts "[INFO][CONST] Not constraining #{name} based on spot/non-spot"
+        pc << {
+          type: 'memberOf',
+          expression: "attribute:capacity-provider.name != on-demand-capacity-provider",
+        }
+        pc << {
+          type: 'memberOf',
+          expression: "attribute:capacity-provider.name != spot-capacity-provider",
+        }
+      end
+    end
+  end
+
+  # https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-placement-strategies.html
+  def _placement_strategy
```

Comment:
figure out a way to show how a service was distributed across AZ and instances

Reply:
Added https://www.pivotaltracker.com/story/show/176578121 as I won't get to this now.